infinitory/errors.py

```diff
@@ -0,0 +1,137 @@
+
+    def load_reports_fast(self, pupdb):
+        # Takes Much Longer
+        self.reports = dict()
+        for report in pupdb.query('reports[] { hash in nodes[latest_report_hash] { } }'):
+            if report['status'] == 'failed':
+                self.reports[report['certname']] = report
+
+    def load_reports(self, pupdb):
+        for report in pupdb.query('nodes[certname, latest_report_hash] { }'):
+            cache_file = "./cache/%s" % report["latest_report_hash"]
+            if os.path.isfile(cache_file):
+                full_report = pickle.load(open(cache_file, "rb"))
+                sys.stdout.write('#')
+            else:
+                full_report = pupdb.query('reports[] { hash = "' + report["latest_report_hash"] + '" }')
+                pickle.dump( full_report, open(cache_file, "wb" ) )
+                sys.stdout.write('.')
+            sys.stdout.flush()
+
+            self._reports[report["certname"]] = full_report[0]
+
+    def stored_errors(self):
+        return [
+            "Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Function Call, Untrusted facts (left) don't match values from certname (right)"
+        ]
+
+    def matches_stored_error(self, message):
+        for se in self.stored_errors():
+            if message.startswith(se):
+                return True, se
+
+        return False, ""
+
+    def set_all_errors(self, input):
+        # Mainly used during testing to mock values
+        self._all_errors = input
+
+    def set_unique_errors(self, input):
+        # Mainly used during testing to mock values
+        self._unique_errors = input
+
+    def all_errors(self):
+        retval = []
+        for i, ue in enumerate(self._all_errors):
+            retval.append({
+                "other": ue
+            })
+        return retval
+
+    def unique_errors(self):
+        retval = []
+        for i, ue in enumerate(self._unique_errors):
+            retval.append({
+                "other": ue
+            })
+        return retval
+
+    def clean_error_message(self, error_message):
+        matched, stored_error = self.matches_stored_error(error_message)
+
+        if matched:
+            return stored_error
+
+        return error_message
+
+    def modify_unique_errors_at(self, i, log_level, hostname):
+
+        new_hostname_list = self._unique_errors[i]['hostnames']
+        new_hostname_list.append(hostname)
+
+        self._unique_errors[i] = {
+            'count': self._unique_errors[i]['count'] + 1,
+            'level': log_level,
+            'hostnames': new_hostname_list,
+            'message': self._unique_errors[i]['message']
+        }
+
+    def append_unique_error(self, error_message, log_level, hostname):
+        for i, ue in enumerate(self._unique_errors):
+            if ue['message'] == error_message:
+                self.modify_unique_errors_at(i, log_level, hostname)
```

Comment:
Use a `dict` for  `_unique_errors`. Or, it would probably be better to use [`defaultdict`](https://docs.python.org/2/library/collections.html#collections.defaultdict).

Reply:
Ahh yes, see above, I wanted to use a dict, but refactored it this way because cellformatter expects a list.